### PR TITLE
add symbol.SwapAxis operator, just can do Forward().

### PR DESCRIPTION
**Pull request by starimpact**
this is a early version, can run now, but haven't been tested.
This is just a look-version, means let you have a look and give me some ideas.

### CODE REVIEW

**Comment by starimpact** [comment]
I am sorry for did not see your comments.

**Comment by starimpact** [comment]
thanks for your suggestions.

**Comment by starimpact** [comment]
I have finished the new code based on your suggestions.
And the Backward is also completed.
Please check the code for me.Give me more suggestions.
Thank you very much.:smile:

File: make/config.mk [hunk 1]
```diff
@@ -38,15 +38,15 @@ ADD_CFLAGS =
 #---------------------------------------------
 
 # whether use CUDA during compile
-USE_CUDA = 0
+USE_CUDA = 1
```
Comment (tqchen) [inline make/config.mk]: Please change this back to default, as most users don't have cuda

File: src/operator/swapaxis-inl.h [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+#include <algorithm>
+#include <map>
+#include <vector>
+#include <string>
+#include <utility>
+#include "./operator_common.h"
+
+#define SWAPAXIS_DBG 0
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: remove debug macro

File: src/operator/swapaxis-inl.h [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  	this->param_ = p;
+  }
+  
+  
+  void Reshape2Five(Shape<5> &inter_shape, TShape &shape, uint32_t dim1, uint32_t dim2)
+  {
+    int ndim_in = shape.ndim();
+    int si;
+    for (si = 0; si < 5; si++)
+    {
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: style: use for() {

File: src/operator/swapaxis-inl.h [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    inter_shape[1] = shape[dim1];
+    //dim_2
+    for (si = dim1+1; si < dim2; si++)
+    {
+      inter_shape[2] *= shape[si];
+    }
+    //dim_3
+    inter_shape[3] = shape[dim2];
+    //dim_4
+    for (si = dim2+1; si < ndim_in; si++)
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: space between dim2 and +1.

File: src/operator/swapaxis-inl.h [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  }
+};
+
+//a1 higher dimension to be swapped, assert a1 > a2
+//a2 lower dimension to be swapped
+template<typename xpu>
+class SwapAxisOp : public Operator {
+ public:
+  explicit SwapAxisOp(SwapAxisParam p) {
+    CHECK_LT(p.dim1, p.dim2) << "dim1 must be lower than dim2.";
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: consider automatically swap if not the case

File: src/operator/swapaxis-inl.h [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    
+    Shape<5> inter_shape2 = inter_shape;
+    inter_shape2[1] = inter_shape[3];
+    inter_shape2[3] = inter_shape[1];
+    
+    Tensor<xpu, 5> inter_data_out = data_out.get_with_shape<xpu, 5, real_t>(inter_shape2, s);
+    
+    TShape shape_tmp = shape_in;
+    dwTmp = shape_tmp[dim1];
+    shape_tmp[dim1] = shape_tmp[dim2];
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: std::swap(shape_tmp[dim1], shap_tmp[dim2])

File: src/operator/swapaxis-inl.h [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    
+#if SWAPAXIS_DBG
+    printf("hello swapaxis Forward!\n");
+#endif
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+
+    __swapaxis(s, in_data, out_data);
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: Use Google code style SwapAxis

File: src/operator/swapaxis-inl.h [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    {
+      std::cout << "Have no input data.\n";
+      return false;
+    }
+    TShape &shape0 = (*in_shape)[SwapAxis::kData];
+#if SWAPAXIS_DBG
+    printf("in_shape_num:%d\n", input_num);
+    printf("in_shape_0, dim:%d, size:%d\n", (int)shape0.ndim(), (int)shape0.Size());
+#endif
+    if (shape0.ndim() !=  4)
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: this restriction is no longer needed

File: src/operator/swapaxis-inl.h [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    return ptr;
+  }
+
+  std::string TypeString() const override {
+    return "SwapAxis";
+  }
+/*
+  std::vector<int> DeclareBackwardDependency(
+    const std::vector<int> &out_grad,
+    const std::vector<int> &in_data,
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: override DeclarebackwardDependency
```
return out_grad[SwapAxis::kOut];
```

File: src/operator/swapaxis-inl.h [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+  explicit SwapAxisOp(SwapAxisParam p) {
+    CHECK_LT(p.dim1, p.dim2) << "dim1 must be lower than dim2.";
+#if SWAPAXIS_DBG
+    printf("hello swapaxis SwapAxisOp:dim1:%d, dim2:%d!\n", p.dim1, p.dim2);
+#endif
+  	this->param_ = p;
+  }
+  
+  
+  void Reshape2Five(Shape<5> &inter_shape, TShape &shape, uint32_t dim1, uint32_t dim2)
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: Use google C style pass non-const ref as pointer, and other as const reference

File: src/operator/swapaxis-inl.h [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    int ndim_in = shape.ndim();
+    int si;
+    for (si = 0; si < 5; si++)
+    {
+      inter_shape[si] = 1;
+    }
+    //dim_0
+    for (si = 0; si < dim1; si++)
+    {
+      inter_shape[0] *= shape[si];
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: consider use http://en.cppreference.com/w/cpp/algorithm/accumulate

File: src/operator/swapaxis-inl.h [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    Shape<5> inter_shape;
+    
+    Reshape2Five(inter_shape, shape_in, dim1, dim2);
+    
+    Tensor<xpu, 5> inter_data_in = data_in.get_with_shape<xpu, 5, real_t>(inter_shape, s);
+    
+    int dwTmp = 0;
+    
+    Shape<5> inter_shape2 = inter_shape;
+    inter_shape2[1] = inter_shape[3];
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: use std::swap

**Comment by tqchen** [comment]
Thanks for the contribution. I have made a few comments on the code, in general

- Code Style, we use Google C++ code style, you can reproduce the linter check locally using
```
make lint
```
- [ ] Try to use standard functions when possible
    - This include std::swap and std::accumulate
- [ ] DeclareBackwardDependency, as Backward only depend on out_grad
   - c.f. http://mxnet.readthedocs.org/en/latest/developer-guide/operator.html
- [ ] Please also write a testcase in https://github.com/dmlc/mxnet/blob/master/tests/python/unittest/test_operator.py

**Comment by starimpact** [comment]
Thanks very much for your suggestions.

**Comment by starimpact** [comment]
function is ready to go. please check it.
maybe there is a little code style problem.
:laughing: :smile: :smile_cat: :smiling_imp:

**Comment by starimpact** [comment]
I come across a problem:
```SwapAxis(..., dim1=2, dim2=3)```can work well on both cpu and gpu.
```SwapAxis(..., dim1=3, dim2=2)```only work well on cpu, but fail on gpu with the following error:
```
INFO:root:Start training with [gpu(0)]
[13:46:12] ./dmlc-core/include/dmlc/logging.h:208: [13:46:12] ./mshadow/mshadow/./tensor_blob.h:530: Check failed: (this->shape_.Size()) == (shape.Size()) TBlob.get_with_shape: new and old shape do not match total elements
[13:46:12] ./dmlc-core/include/dmlc/logging.h:208: [13:46:12] src/engine/./threaded_engine.h:295: [13:46:12] ./mshadow/mshadow/./tensor_blob.h:530: Check failed: (this->shape_.Size()) == (shape.Size()) TBlob.get_with_shape: new and old shape do not match total elements
terminate called after throwing an instance of 'dmlc::Error'
  what():  [13:46:12] src/engine/./threaded_engine.h:295: [13:46:12] ./mshadow/mshadow/./tensor_blob.h:530: Check failed: (this->shape_.Size()) == (shape.Size()) TBlob.get_with_shape: new and old shape do not match total elements
Aborted (core dumped)
```

**Comment by starimpact** [comment]
std::accumulate can not be recognized by nvcc.

**Comment by tqchen** [comment]
Oh, yap. Maybe create our own version of prod function is easier. Then I think it is OK. Thanks

**Comment by tqchen** [comment]
For the error, as it indicates, it means the shape's size do not match the size of TBlob. Likely due to shape initialization error, either InferShape or Shape2Five

**Comment by tqchen** [comment]
The platform dependency issue was likely due to some uninitialized memory(variable) that causes uncertainties, but just my guess

**Comment by starimpact** [comment]
I recompiled the all, removed the std::accumulate, and then the problem of ```dim1=3, dim2=2``` of gpu is gone...............

**Comment by tqchen** [comment]
Please update the code and remove debug clause, rebase to the master and check the codestyle by 
```
make lint
```

If the testcase pass, I will do another round of code review. Thanks!

File: make/config.mk [hunk 13]
```diff
@@ -43,10 +43,10 @@ USE_CUDA = 0
 # add the path to CUDA libary to link and compile flag
 # if you have already add them to enviroment variable, leave it as NONE
 # USE_CUDA_PATH = /usr/local/cuda
-USE_CUDA_PATH = NONE
```
Comment (tqchen) [inline make/config.mk]: Please do not change config.mk, copy it to root, and change it on project root

File: src/operator/swapaxis-inl.h [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+  explicit SwapAxisOp(SwapAxisParam p) {
+    CHECK_NE(p.dim1, p.dim2) << "dim1 can not be equal dim2.";
+    if (p.dim1 > p.dim2) {
+      std::swap(p.dim1, p.dim2);
+    }
+  	this->param_ = p;
+  }
+  
+  void Reshape2Five(Shape<5> &inter_shape, const TShape &shape, const uint32_t &dim1, const uint32_t &dim2) {
+    int ndim_in = shape.ndim();
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: index_t

**Comment by starimpact** [comment]
I have no cpplint module.

File: src/operator/swapaxis-inl.h [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    
+    //dim_3
+    inter_shape[3] = shdata[dim2];
+    
+    //dim_4
+    inter_shape[4] = std::accumulate(shdata + dim2 + 1, shdata + ndim_in, 1, std::multiplies<int>());
+    
+  }
+  
+  void __swapaxis(Stream<xpu> *s, const std::vector<TBlob> &in_data, const std::vector<TBlob> &out_data) {
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: Use google code style, SwapAxis

Comment (tqchen) [inline src/operator/swapaxis-inl.h]: This function can be static

File: src/operator/swapaxis-inl.h (hunk 14 above)
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: This function can be static

File: src/operator/swapaxis-inl.h [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+ public:
+  explicit SwapAxisOp(SwapAxisParam p) {
+    CHECK_NE(p.dim1, p.dim2) << "dim1 can not be equal dim2.";
+    if (p.dim1 > p.dim2) {
+      std::swap(p.dim1, p.dim2);
+    }
+  	this->param_ = p;
+  }
+  
+  void Reshape2Five(Shape<5> &inter_shape, const TShape &shape, const uint32_t &dim1, const uint32_t &dim2) {
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: Pass non-const ref as pointer

File: tests/python/unittest/test_operator.py [hunk 17]
```diff
@@ -171,8 +171,29 @@ def test_regression():
...
+	shape = (2, 3, 4)
+	data_tmp = np.ones(shape)
+	data_tmp[0] = 1
+	data_tmp[1] = 2
+	arr_data = mx.nd.array(data_tmp)
+	swap0 = mx.symbol.SwapAxis(data=data, dim1=0, dim2=2)
+	swap = mx.symbol.SwapAxis(data=swap0, dim1=1, dim2=2)
+	exe_c = swap.bind(mx.cpu(), args=[arr_data])
+	exe_c.forward()
+	out = exe_c.outputs[0].asnumpy()
```
Comment (tqchen) [inline tests/python/unittest/test_operator.py]: compare out with swapaxes in numpy function, do assert as in other testcases

File: tests/python/unittest/test_operator.py [hunk 18]
```diff
@@ -171,8 +171,29 @@ def test_regression():
                      lambda x: x,
                      lambda x, y : x - y)
 
+def test_SwapAxis():
```
Comment (tqchen) [inline tests/python/unittest/test_operator.py]: rename to test_swapaxes

**Comment by tqchen** [comment]
You can insall cpplint and pylint by 
```
pip install cpplint pylint
```

**Comment by starimpact** [comment]
can you help me for the following problem?
```python
def test2():
    data_in = mx.symbol.Variable('data')
    conv = mx.symbol.Convolution(data=data_in, kernel=(3, 3), num_filter=16)
    datatmp = np.ones((1, 1, 32, 64))
    mxdata = mx.nd.array(datatmp)
    weightmp = np.ones((16, 1, 3, 3))
    mxweight = mx.nd.array(weightmp)
    biastmp = np.zeros(16)
    mxbias = mx.nd.array(biastmp)
    exe_c = conv.bind(ctx=mx.gpu(0), args=[mxdata, mxweight, mxbias])
    out = exe_c.forward()
    print out

test2()
```
Error
```
None
[[14:29:51] ./dmlc-core/include/dmlc/logging.h:208: [14:29:51] ./mshadow/mshadow/./stream_gpu-inl.h:98: Check failed: (err) == (CUBLAS_STATUS_SUCCESS) Create cublas handle failed
14:29:51] ./dmlc-core/include/dmlc/logging.h:208: [14:29:51] ./mshadow/mshadow/./stream_gpu-inl.h:98: Check failed: (err) == (CUBLAS_STATUS_SUCCESS) Create cublas handle failed
terminate called recursively
terminate called after throwing an instance of 'Aborted (core dumped)
```

**Comment by starimpact** [comment]
what's the next step?
```
=====69/70 cpp-header files passed check=====
src/operator/swapaxis-inl.h: 34 Errors of 4 Categories map={'whitespace': 29, 'runtime': 2, 'readability': 2, 'build': 1}
=====57/58 cpp-soruce files passed check=====
src/operator/swapaxis.cc: 2 Errors of 1 Categories map={'readability': 2}
```

**Comment by tqchen** [comment]
resolve all the comments by cpplint, there are detailed messages

**Comment by tqchen** [comment]
For output, we need to do
```
exec_c.forward()
out = exec.output[0].asnumpy()
print out
```

The error was due to we did not call asnumpy to wait for the result, and the system start to shutdown before the computation starts.

**Comment by starimpact** [comment]
I can see nothing comments in the swapaxis files by cpplint? where are they?

**Comment by starimpact** [comment]
there is something error...
```
=====69/70 cpp-header files passed check=====
src/operator/swapaxis-inl.h: 34 Errors of 4 Categories map={'whitespace': 29, 'runtime': 2, 'readability': 2, 'build': 1}
=====57/58 cpp-soruce files passed check=====
src/operator/swapaxis.cc: 2 Errors of 1 Categories map={'readability': 2}
=====40/40 python files passed check=====
2 files failed lint
make: *** [lint] Error 1
```

**Comment by tqchen** [comment]
Hmm the error message occurs before these when it scan through files. You
may need to scroll up a bit to see it

**Comment by starimpact** [comment]
cpu is work right now, but not for gpu:
```python
def test2():
    data_in = mx.symbol.Variable('data')
    conv = mx.symbol.Convolution(data=data_in, kernel=(3, 3), num_filter=16)
    datatmp = np.ones((1, 1, 32, 64))
    mxdata = mx.nd.array(datatmp)
    weightmp = np.ones((16, 1, 3, 3))
    mxweight = mx.nd.array(weightmp)
    biastmp = np.zeros(16)
    mxbias = mx.nd.array(biastmp)
    exe_c = conv.bind(ctx=mx.gpu(0), args=[mxdata, mxweight, mxbias])
    exe_c.forward()
    out = exe_c.outputs[0].asnumpy()
    print out

test2()
```
Error
```
[14:50:16] ./dmlc-core/include/dmlc/logging.h:208: [14:50:16] ./mshadow/mshadow/./tensor_blob.h:508: Check failed: Device::kDevMask == dev_mask_ && DataType<DType>::kFlag == type_flag_ TBlob.get: device type do not match specified type
[14:50:16] ./dmlc-core/include/dmlc/logging.h:208: [14:50:16] src/engine/./threaded_engine.h:295: [14:50:16] ./mshadow/mshadow/./tensor_blob.h:508: Check failed: Device::kDevMask == dev_mask_ && DataType<DType>::kFlag == type_flag_ TBlob.get: device type do not match specified type
terminate called after throwing an instance of 'dmlc::Error'
  what():  [14:50:16] src/engine/./threaded_engine.h:295: [14:50:16] ./mshadow/mshadow/./tensor_blob.h:508: Check failed: Device::kDevMask == dev_mask_ && DataType<DType>::kFlag == type_flag_ TBlob.get: device type do not match specified type
Aborted (core dumped)

```

**Comment by starimpact** [comment]
```make lint ``` is all passed!

File: tests/python/unittest/test_operator.py (hunk 17 above)
Comment (starimpact) [inline tests/python/unittest/test_operator.py]: I am not very clear what your mean.

File: src/operator/swapaxis-inl.h [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+
+  std::map<std::string, std::string> GetParams() const override {
+    return param_.__DICT__();
+  }
+
+  bool InferShape(std::vector<TShape> *in_shape,
+                  std::vector<TShape> *out_shape,
+                  std::vector<TShape> *aux_shape) const override {
+    int input_num = in_shape->size();
+    if (input_num == 0) {
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: CHECK_EQ(input_shape->size(), 1);

File: src/operator/swapaxis-inl.h [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+  std::vector<int> DeclareBackwardDependency(
+    const std::vector<int> &out_grad,
+    const std::vector<int> &in_data,
+    const std::vector<int> &out_data) const override {
+    return {out_grad[SwapAxis::kOut]};
+  };
+/*
+  std::vector<ResourceRequest> ForwardResource(
+      const std::vector<TShape> &in_shape) const override;
+
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: remove the commented fields

File: tests/python/unittest/test_operator.py [hunk 21]
```diff
@@ -171,8 +171,29 @@ def test_regression():
...
+	  data_tmp[0] = 1
+	  data_tmp[1] = 2
+	  arr_data = mx.nd.array(data_tmp)
+	  swap0 = mx.symbol.SwapAxis(data=data, dim1=0, dim2=2)
+	  swap = mx.symbol.SwapAxis(data=swap0, dim1=1, dim2=2)
+	  exe_c = swap.bind(mx.cpu(), args=[arr_data])
+	  exe_c.forward()
+	  out = exe_c.outputs[0].asnumpy()
+	  print data_tmp.shape
+	  print data_tmp
```
Comment (tqchen) [inline tests/python/unittest/test_operator.py]: Do assert, and compare with numpy result, testcases should be able to fail when we run it and something wrong happens. We cannot rely on eyebow to detect errors.

Something like

```
assert same(out, np.swapaxes(data_tmp, dim1, dim2))
```

File: src/operator/swapaxis-inl.h [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+  }
+
+  virtual void Forward(const OpContext &ctx,
+                       const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+
+    __swapaxis(s, in_data, out_data);
```
Comment (tqchen) [inline src/operator/swapaxis-inl.h]: rename _swapaxis to SwapAxis

**Comment by tqchen** [comment]
Thanks for the good job! I have last few comments, please address them and rebase to resolve the conflict to current master here.

http://mxnet.readthedocs.org/en/latest/contribute.html#how-to-resolve-conflict-with-master

The conflict might have something to do with commits you have in mshadow, if that is the case, the best way might be reset mshadow's version, or keep a copy of your files, and do a clean fork

**Comment by starimpact** [comment]
Done, please have a check!

**Comment by starimpact** [comment]
oh, sorry, have not updated your newest comments.

**Comment by starimpact** [comment]
I think you forgot I can not push to your repository.

**Comment by starimpact** [comment]
Do I have another way to add your newest commits without do second time fork?

**Comment by tqchen** [comment]
You can do a rebase http://mxnet.readthedocs.org/en/latest/contribute.html#how-to-resolve-conflict-with-master

**Comment by starimpact** [comment]
something is wrong when I do rebase your mxnet.

**Comment by starimpact** [comment]
I do fetch your mxnet, but my repository is changed to my older version one when I do rebase.
really strange.

**Comment by starimpact** [comment]
I have pushed the newest code to my forked mxnet, can you check it?

**Comment by starimpact** [comment]
I need a little time to figure out the strange problem of rebase.

**Comment by starimpact** [comment]
Can I do merge?

**Comment by tqchen** [comment]
The general instruction is here http://mxnet.readthedocs.org/en/latest/contribute.html#how-to-resolve-conflict-with-master 

If you find files with conflicts, edit the files and merge the conflict, and do a git add as indicated in the instruction

**Comment by starimpact** [comment]
the situation is: I have commit a lot in my local repository, when I rebase to your newest master, my current work directory content is updated to the first commit. What can I do?

**Comment by starimpact** [comment]
OK, likely I found the way....

**Comment by starimpact** [comment]
ok ,now, please check.

**Comment by tqchen** [comment]
closed due to #519

**Pull request closed without merging**